### PR TITLE
Contribute guidelines: amend documentation for policies around modifying others' work

### PR DESCRIPTION
**Pull request by ioannisg**
should be reflecting the feedback received in the process improvement meetings

### CODE REVIEW

File: doc/contribute/modifying_contributions.rst
Comment (pabigot) [inline doc/contribute/modifying_contributions.rst:53]: The criteria and process for characterizing a PR as *stale* do not appear to be defined.

Also I sometimes do this on non-stale PRs when then original author agrees to let me, generally because of differential git skills and the complexity of getting the exact desired solution in place with a clean commit history.

File: doc/contribute/index.rst
Comment (pabigot) [inline doc/contribute/index.rst:141]: A real-world application of this is #25971.

The first commit contains some fragments that were unchanged from the [downstream commit](https://github.com/nrfconnect/sdk-zephyr/pull/305/commits/5500a99cc02a1675663a9126bbd71e5c248c750c), but the bulk of the material is new.    I kept the SOB line from @sigvartmh to acknowledge that fragments were taken unmodified (so asserting the origin for what I kept is valid), but submitted it under my authorship---in part because I completely rewrote the commit message and felt that implying he had primary responsibility was inappropriate.

I'm willing to engage in discussion about whether this was the best way to proceed.  AFAIK it was found acceptable by the original author.

File: doc/contribute/modifying_contributions.rst
Comment (pabigot) [inline doc/contribute/modifying_contributions.rst:61]: s/indentify/identify/

The criteria and process for characterizing an original author as *unresponsive* are not defined.

Also I would propose that if any material from the original is retained the signed-off-by line of the original should be kept, based on [this discussion](https://stackoverflow.com/a/1962112/236848) of the intent behind that statement (chain-of-responsibility).

File: doc/contribute/modifying_contributions.rst
Comment (pabigot) [inline doc/contribute/modifying_contributions.rst:17]: Generally paragraphs that precede a bullet list in rst text in Zephyr end with either `:` or `::`.  Having neither doesn't seem right.

**Review by pabigot — COMMENTED** [review]
A few points of perspective (and a couple nits) since this sort of thing comes up quite frequently IMO.

File: doc/contribute/index.rst
Comment (ioannisg) [inline doc/contribute/index.rst:136]: Trying to capture @galak 's input here: adding: "Any contributions made as part of submitted pull requests are considered free for the Project to use."

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:53]: Agreed, this needs to be defined somewhere.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [inline doc/contribute/modifying_contributions.rst:57]: Simplified (based on @galak 's fb)

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [inline doc/contribute/modifying_contributions.rst:64]: Adding a comment about the PR number as an example of how an author can link to someone else's work

File: doc/contribute/index.rst
Comment (nandojve) [reply doc/contribute/index.rst:143]: Small comment about this. When we cherry-picks any commit and keep it untouched we know who made the contribution. When we change anything, we don't know who made what. I think this could hurt people and create unnecessary discussions and worst, people stop send contributions. If a cherry-picks is important we need keep as original (even with bugs). If we need change a work, it would be cherry-picked without any modifications and then create another commit with the changes and say why. That makes things clear instead blurry.

Comment (ioannisg) [reply doc/contribute/index.rst:143]: @nandojve thanks for your comment: I think your concern is reflected in the document. To summarize
- cherry-picking without content change is allowed and should leave commit  author/sign-off as is
- pick & modify leaves 2 options
   - reach out to original author to ACK
   - send the patch with own author/sign-off and leave a note about source of inspiration (optional)

> it would be cherry-picked without any modifications and then create another commit with the changes and say why.

@nandojve this was actually discussed and , although it is a valid option, it is not always ideal, since this creates extra noise in PR reviews but, most importantly, it affects bisectability (in case it contains obvious bugs)

**Review by jfischer-no — APPROVED** [review]

File: doc/contribute/index.rst
Comment (nandojve) [reply doc/contribute/index.rst:143]: > pick & modify leaves 2 options
> * reach out to original author to ACK
> * send the patch with own author/sign-off and leave a note about source of inspiration (optional)

Small changes like adjusts to fit: add below is enough in my opinion because it's a clean reason and we need keep tracking of what happened somehow.
```
[Project Maintainer: Renamed test methods according to naming convention.]
Signed-off-by: Project Maintainer <project.maintainer@example.com>
```

The huge changes that imply own the code we **must** communicate original Author not "should" as mentioned. It seems section **Recommended policies** is not useful because say that anyone can get anything and do whatever they want without say anything to anyone. One thing is the end object zephyr project it self, the other thing is the sequence (commits) that makes the zephyr project. There is no Apache Licence on every commit. So, when someone takes code from others and do a lot of changes it must communicate, it must explain the why, it should invite to review, it should ask Author ack. We shouldn't give super powers to anyone, we should propose a contract that is fair to every one.

Take this for example:
> Any contributions made as part of submitted pull requests are considered free

True, as is. Above don't give rights to change the inside content of a commit.

I hope above not sounds like a complain. My point is, we need think there are people sending code, their deserves respect and they have feelings. If Zephyr don't want to be a dictator it must have policy to avoid that.

**Review by nashif — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Comment by pabigot** [comment]
I'm just commenting because I'm not in the process group or on TSC and this is project-level policy, but:
* Typo is still not fixed;
* Formatting inconsistency with rest of docs is not fixed;
* Current content suggests it's OK to integrate and adapt work from other people without preserving their signed-off-by line, thus losing the [https://stackoverflow.com/a/1962112/236848](chain-of-responsibility) that confirms the material conforms to the license, copyright, and authorship it proposes.

I don't quite go as far as @nandojve in saying an adopted commit can't be modified if it's clearly something that in review would be rejected unless the commit was changed (i.e., squashing a cleanup/fix).  But I do agree it is inappropriate to erase in-code/commit acknowledgement of the original author.  A link and comment in a PR doesn't do it.

Under the process as currently documented, I will go out of my way to reimplement work that I could otherwise adapt, if retaining blocks of unmodified code acknowledged through the original author's SOB is not an option.

Perhaps a section should be added that indicates exactly why Zephyr requires a signed-off-by line, and what presence of that line is intended to convey.  Is it "X approves of this content", "X contributed to this content", or something else?

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst]: 1. Typo is now fixed.
2. unresponsive authors: yeap, this needs to be defined explicitly, but it is not part of this change, so I prefer to address it outside this PR

>Also I would propose that if any material from the original is retained the signed-off-by line of the original should be kept, based on this discussion of the intent behind that statement (chain-of-responsibility).

We discussed this extensively in the Process Forum (participation is open to all, btw). Retaining the original sign-off cannot be required in all cases; this is exactly the problem we are trying to solve: picking and heavily modifying commits of others, while retaining original sign-off. We don't want to do that, unless the original authors ACKed this.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:53]: @pabigot stale PRs and ISsues are defined here:
https://docs.zephyrproject.org/latest/development_process/dev_env_and_tools.html

I will add a link.

Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst]: > Also I sometimes do this on non-stale PRs when then original author agrees to let me, generally because of differential git skills and the complexity of getting the exact desired solution in place with a clean commit history.

Agreed. I believe, the way I wrote the sentence still allows you to continue doing this:
> should **try to limit** the above practice

But it is not encouraged as a long-term solution. Agreeing with original author is something we refer to later (for modifying content) but in theory it could also apply here, so I think we are sort of covered.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst]: Agreed. Fixed this now

**Comment by ioannisg** [comment]
> I'm just commenting because I'm not in the process group or on TSC and this is project-level policy, but:
> 
> * Typo is still not fixed;
> * Formatting inconsistency with rest of docs is not fixed;
> * Current content suggests it's OK to integrate and adapt work from other people without preserving their signed-off-by line, thus losing the [chain-of-responsibility](https://stackoverflow.com/a/1962112/236848) that confirms the material conforms to the license, copyright, and authorship it proposes.
> 
> I don't quite go as far as @nandojve in saying an adopted commit can't be modified if it's clearly something that in review would be rejected unless the commit was changed (i.e., squashing a cleanup/fix). But I do agree it is inappropriate to erase in-code/commit acknowledgement of the original author. A link and comment in a PR doesn't do it.
> 
> Under the process as currently documented, I will go out of my way to reimplement work that I could otherwise adapt, if retaining blocks of unmodified code acknowledged through the original author's SOB is not an option.
> 
> Perhaps a section should be added that indicates exactly why Zephyr requires a signed-off-by line, and what presence of that line is intended to convey. Is it "X approves of this content", "X contributed to this content", or something else?

@pabigot 

Typo, formatting and link to definition of stale work are now fixed/added.

> But I do agree it is inappropriate to erase in-code/commit acknowledgement of the original author. A link and comment in a PR doesn't do it.

Agreed; this is captured and addressed in two points that are made here:
1. We state that reviewers are encouraged to help out in PRs rather than picking original work into their own PRs.
2. There is always the possibility for you to get an ACK from the original author and retain their sign-off, but we should also have a workaround, if the author cannot ACK.

**Comment by pabigot** [comment]
> Agreed it sounds inappropriate but we need to ensure development can progress smoothly.
> This is captured and addressed in two points that are made in this section:
> 
> 1. We state that reviewers are encouraged to help out in PRs rather than picking original work into their own PRs.
> 2. There is always the possibility for you to get an ACK from the original author and retain their sign-off, but we should also have a workaround, if the author cannot ACK. In that case, I do not think it is inappropriate to have a comment/link to the original work as the reference.

Then IMO you need to document that in Zephyr the signed-off-by line means that the named person **approves** of the content of the commit, rather than provided material that is present in the commit as it indicates in Linux.  Without that it's unclear that a maintainer or contributor that tries to use somebody else's work can remove the SOB in the case of an unresponsive author of the original commit.

https://stackoverflow.com/a/14044024/236848 proposes a multi-level sign-off which might be a better approach than what we're getting here.

File: doc/contribute/index.rst
Comment (ioannisg) [reply doc/contribute/index.rst:143]: >The huge changes that imply own the code we must communicate original Author not "should" as mentioned. It seems section Recommended policies is not useful because say that anyone can get anything and do whatever they want without say anything to anyone.

> So, when someone takes code from others and do a lot of changes it must communicate, it must explain the why, it should invite to review, it should ask Author ack. We shouldn't give super powers to anyone, we should propose a contract that is fair to every one.

@nandojve thanks for stressing this.
So, I think all this is present in the text; I believe you propose to make this more of a requirement, not a recommendation. So I am thinking of
- changing "Recommended policies" to "Accepted policies"
- changing **should** to **shall** in:
    - explain why
    - invite original author to the review
    - ask for ACK from them

Would these changes satisfy your concerns here?

Comment (nandojve) [reply doc/contribute/index.rst:143]: @ioannisg yes! I believe most important is explain and communicate. With that, no one have space to complain. The invide is the best way to communicate because shows that Zephyr cares about what is going on. It is up to Author ACK and participate on the review process. However, we need care to give time to Author manifest. This should follow normal review process and it may be important reinforce that.

Comment (ioannisg) [reply doc/contribute/index.rst:143]: that's all fine, @nandojve , I did the proposed updates, please, take a look.

**Comment by ioannisg** [comment]
> Then IMO you need to document that in Zephyr the signed-off-by line means that the named person approves of the content of the commit, rather than provided material that is present in the commit as it indicates in Linux. Without that it's unclear that a maintainer or contributor that tries to use somebody else's work can remove the SOB in the case of an unresponsive author of the original commit.

While this is an option, isn't this issue covered by the text I 've added:
```
Any contributions made as part of submitted pull requests are considered free
for the Project to use.
```

under the "Developer Certification of Origin (DCO)" section?

**Comment by pabigot** [comment]
> While this is an option, isn't this issue covered by the text I 've added:

Yes. No. Maybe. It depends.  Here's the text that concerns me:

>* submit the modified patches as their *own* work (i.e. with their
>  *own* sign-off line and author identity). In this case, the developer
>  may want to identify in the commit message(s) the original source the
>  submitted work is based on (mentioning, for example, the original PR
>  number).

Is the intent of this to indicate that the modified commit should **not** retain the original sign-off?

I kinda have a problem with that, as the burden of defending against license violations subsequently revealed to be present in the original contribution is much harder to pass back to that contributor, who is no longer associated with it in the historical record of what was done.  (git repositories last forever; PRs and commits are more transient.)

That can be clarified if Zephyr policy makes explicit what the presence of a sign-off means in Zephyr: contributed parts of, or approves of, or something else.  If it's "contributed", then the sign-off should remain even if the original contributor vehemently objects to the rework.  They accepted that as a possibility when they signed-off their original contribution.  Only if somebody can make a case that nothing in the commit is derived from a different contribution should a sign-off be removed.  IMO.

**Review by nandojve — APPROVED** [review]

File: doc/contribute/index.rst
Comment (carlescufi) [inline doc/contribute/index.rst:141]: I would suggest:

* the content of the patches will not be substantially modified,
* the cherry-picked commits or portions of a commit shall preserve the original sign-off messages and
   the author identity.

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:19]: cherry-picks

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:26]:
```suggestion
* a developer pushes to a branch or pull request opened by another
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:29]:
```suggestion
  * assist in updating pull requests in order to get the patches merged
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:31]:
```suggestion
  * drive stale pull requests to completion so they can be merged
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:37]:
```suggestion
A developer who intends to cherry-pick and potentially modify patches sent by
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:43]:
```suggestion
* invite the original author of the patches to their pull request review.
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:45]:
```suggestion
A developer who intends to force-push to a branch or pull request of
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:48]:
```suggestion
is done to drive the pull request review  to completion when the pull
```

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:63]:
```suggestion
  shall identify in the commit message(s) the original source the
```
I think we should always mention where the original work comes from

File: doc/contribute/modifying_contributions.rst
Comment (carlescufi) [inline doc/contribute/modifying_contributions.rst:68]:
```suggestion
  Contributors should be uncheck the box *“Allow Edits By Maintainers"*
```

**Review by carlescufi — DISMISSED** [review]
Great work, @ioannisg. I have a few edits but this is looking very good.

**Comment by carlescufi** [comment]
> I kinda have a problem with that, as the burden of defending against license violations subsequently revealed to be present in the original contribution is much harder to pass back to that contributor, who is no longer associated with it in the historical record of what was done. (git repositories last forever; PRs and commits are more transient.)

I think the middle-ground here could be to add a note specifying that the commit message in this case must **always** refer to the original source of the work. That way we can always trace it back to the original signed-off line if contention regarding license or copyright arises.

**Review by ioannisg — COMMENTED** [review]
Thanks for the review @carlescufi , I am going to address the feedback. @dleach02 @galak would be nice if you could also give your comments, if any, so I can do a single revision with all the received feedback!

File: doc/contribute/index.rst
Comment (dleach02) [reply doc/contribute/index.rst:136]: do we need to differentiate between DNM PRs where maybe the developer is asking for feedback?

File: doc/contribute/modifying_contributions.rst
Comment (dleach02) [inline doc/contribute/modifying_contributions.rst:9]: Do you need "In general,"?

File: doc/contribute/modifying_contributions.rst
Comment (dleach02) [inline doc/contribute/modifying_contributions.rst:42]: why instead of?

File: doc/contribute/index.rst
Comment (dleach02) [reply doc/contribute/index.rst:143]: It seems like the easiest way to deal with cherry-picking is to cherry-pick the changes as standalone commits in your PR then add your changes on top of them as separate commits where your commit message will identify why you made changes (and maybe why you cherry-picked a different PR instead of helping to push that PR to merge).

File: doc/contribute/index.rst
Comment (ioannisg) [reply doc/contribute/index.rst:136]: @dleach02 I don't think so; and this was not brought up in the meetings as far as I remember. (Also not everyone can add a label to a PR such as DNM, RFC, only collaborators can do that.) If you see a reason to differentiate, pls elaborate.

File: doc/contribute/index.rst
Comment (ioannisg) [reply doc/contribute/index.rst:141]: Yes, this is good improvement, will apply it.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:19]: yeap, thanks

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:26]: done, thanks

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:29]: thanks, done.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:31]: yeap, that's fine , too. thanks

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:37]: done, too.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:43]: fixed

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:45]: fixed

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:48]: thanks, done

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:63]: @dleach02 @carlescufi no objections to this.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:68]: thanks. I suppose it should be "should uncheck"

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:9]: Well, no. Thanks, I am removing this.

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:42]: Hm, I guess "instead of", here, is used to distinguish between "doing that" and "assisting in getting..." 
As it seems not to be clear, I'll replace "doing that" with the actual verb action :)

Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:42]: fixed now

**Comment by ioannisg** [comment]
>It seems like the easiest way to deal with cherry-picking is to cherry-pick the changes as standalone commits in your PR then add your changes on top of them as separate commits where your commit message will identify why you made changes (and maybe why you cherry-picked a different PR instead of helping to push that PR to merge).

@dleach02 the reason not to strictly enforce the principle of fixup commits on top of originals, is that these standalone commits might be "breaking" master, which will be a problem with bisectability. 

> and maybe why you cherry-picked a different PR instead of helping to push that PR to merge)

This is covered. (We force the author to comment in the PR why cherry-picking was done instead of helping out original review)

**Comment by ioannisg** [comment]
> > I kinda have a problem with that, as the burden of defending against license violations subsequently revealed to be present in the original contribution is much harder to pass back to that contributor, who is no longer associated with it in the historical record of what was done. (git repositories last forever; PRs and commits are more transient.)
> 
> I think the middle-ground here could be to add a note specifying that the commit message in this case must **always** refer to the original source of the work. That way we can always trace it back to the original signed-off line if contention regarding license or copyright arises.

@pabigot @carlescufi this is done now. I re-wrote this to stress that the commit shall always refer to the original source of contribution.

**Comment by ioannisg** [comment]
@dleach02 @pabigot your feedback is now addressed

**Review by dleach02 — APPROVED** [review]

File: doc/contribute/index.rst
Comment (stephanosio) [reply doc/contribute/index.rst:136]: Not DNM, but maybe draft (IIRC, everyone can do this)?

Comment (pabigot) [reply doc/contribute/index.rst:136]: I think a case could be made that if it's a draft PR it's not a submitted pull request, so no it's not free for the project to use.  But since "ready" PRs can be converted back to "draft" there's a problem where permission could be revoked after a commit had been cherry-picked into another PR.

The only case I think is clear is that if a commit doesn't have an SOB line it can't be adopted, regardless of what sort of PR it's in.

Comment (ioannisg) [reply doc/contribute/index.rst:136]: @pabigot is right on PR status - they can switch back and forth, so this won't work.

Comment (ioannisg) [reply doc/contribute/index.rst:136]: > The only case I think is clear is that if a commit doesn't have an SOB line it can't be adopted, regardless of what sort of PR it's in.

Yea, and I believe this is already covered. When cherry-picking we need to preserve the SOB. If there's no SOB, we don't cherry pick.

File: doc/contribute/modifying_contributions.rst
Comment (mnkp) [inline doc/contribute/modifying_contributions.rst:46]: Should we state that "Zephyr contributor shall clarify" _before_ doing the force-push? I.e. a user who wants to force push should first notify the author of the original PR of their intent? Just to avoid a situation when someone, suddenly, makes a force push while the original author has already made some changes localy but didn't push their work to the github repo. I assume that's the expected behavior but at the moment it is not explicit.

**Review by MaureenHelm — APPROVED** [review]

**Review by galak — APPROVED** [review]

**Review by ruuddw — APPROVED** [review]

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:46]: @mnkp thanks for your comment: I would think this is rather intuitive; we should always be careful with force-pushing, that's why I think this might be not needed here. Besides:
- we stress that force-pushing to other's branch/PR should be limited to stale PRs, i.e. with un-responsive authors, and for reasons like to drive the PR to completion (when the author is not able to do themselves), so I would say the "race-condition" is not the use-case here :)

Comment (pabigot) [reply doc/contribute/modifying_contributions.rst:46]: "Race-condition" is not the point.  The point is who has ownership of the *branch* and so should be in control of what's in it.  The branch is in the submitter's repository.  Anybody pushing to a branch in my repos without getting a positive acknowledgement from me that it's OK is not going to be well thought of.  People may manage their branches in ways we don't understand, and a force push from other content might break that management.

My personal policy is in line with that: I'll only push if the owner of the repo authorizes it.  In any other case where I was taking over a stale PR under the conditions authorized in this policy I would issue a new PR from my repo, referencing the original which would be closed.

**Review by kestewart — APPROVED** [review]

File: doc/contribute/modifying_contributions.rst
Comment (ioannisg) [reply doc/contribute/modifying_contributions.rst:46]: > My personal policy is in line with that: I'll only push if the owner of the repo authorizes it. In any other case where I was taking over a stale PR under the conditions authorized in this policy I would issue a new PR from my repo, referencing the original which would be closed.

@pabigot although I am fully supporting more transparency and more communication in the Project (and your proposal is ideal with respect to transparency/communication), I am afraid, this may introduce more problems than what it solves.
- asking for authorization to force push: then we need to define how to do it: Slack? E-mail? GH comment? And for how long we should wait until we flag the author as non-responsive? 
- issuing a new PR from your repo is problematic: you lose comment/review history, and will need to ask around, once more, for all the approvals you had in the original PR. IMHO, this slows down a lot the process.

I assume a common case of a PR that is already reviewed, approved, but with Git conflicts, or broken CI, and the author being on holiday for a while; a trivial force-push should do it (with proper commenting, of course, as we explain here) and for the Release Engineer it would be much easier to get the PR merged, rather than looking at a brand-new PR with the same changes.

If anyone force-pushes to your branch to get CI/conflicts fixed and get it merged; I don't clearly see why this should, in general not be well thought of.

In any case, authors can always uncheck the relevant GH box to disallow adding / modifying commits

**Pull request merged**
